frontend/src/helpers/pdf.js

```diff
@@ -0,0 +1,46 @@
+import { daysInWeek } from '@/constants/day';
+import { generateColor } from './color';
+import { maxHoursPerDay } from './count';
+
+export function getDocDefinition(timetable, getCardPrimaryText, getCardSecondaryText, monochromeMode) {
+  return {
+    content: Object.keys(timetable).map((timetableTitle, i) => {
+      const timetableData = timetable[timetableTitle];
+      const hoursPerDay = maxHoursPerDay(timetableData);
+      return [
+        { text: timetableTitle, fontSize: 28, bold: true, margin: [0, 0, 0, 10], pageBreak: i === 0 ? null : 'before' },
+        {
+          table: {
+            widths: [30, '*', '*', '*', '*', '*'],
+            body: [
+              [
+                { text: '#', alignment: 'center', bold: true },
+                ...daysInWeek.map(day => ({ text: day, bold: true, margin: [3, 0] })),
+              ],
+              ...Array(hoursPerDay)
+                .fill()
+                .map((_, hourIndex) => [
+                  { stack: ['\n', { text: hourIndex + 1, bold: true, alignment: 'center' }, '\n'], margin: [3, 4] },
+                  ...Array(daysInWeek.length)
+                    .fill()
+                    .map((_, dayIndex) => {
+                      if (!timetableData[dayIndex][hourIndex]) {
+                        return { stack: [] };
+                      }
+                      const subject = JSON.parse(timetableData[dayIndex][hourIndex]);
+                      return {
+                        stack: [{ text: getCardPrimaryText(subject), bold: true }, '\n', getCardSecondaryText(subject)],
+                        fillColor: monochromeMode ? null : generateColor(getCardPrimaryText(subject)),
+                        fillOpacity: 0.6,
+                        margin: [3, 4],
+                      };
+                    }),
+                ]),
+            ],
+          },
+        },
+      ];
+    }),
+    pageOrientation: 'landscape',
+  };
+}
```

Comment:
This code is very difficult to read and maneuver around. A lot of nested code, basically single-line code spread out over multiple lines. Try to write code for whoever is reading this in the future; be it other people or yourself. While writing it take into account how easy it is to review, how easy it will be for somebody to change it, etc. Regardless if it were a critical part of the app like an algorithm or one that is less so like a PDF export. Most things you write in the future will be maintained by somebody. This app is not one of those (and it is late in the game) but the intention with it was to recreate such environments and for me to point out potential pitfalls. So that is what I am doing. I do not know if this is your usual approach to writing code but my advice is for the future to avoid this approach completely. I personally look at writing readable code as a professional courtesy towards my colleagues (including myself in that 😄). Most of my time is spent reading code and most other people as well. The nature of the job and working on projects that last longer than the first push to production.

Below is a version I consider to be more readable but not the "ultimate" form of it since there is not much time left.
```js
import { daysInWeek } from '@/constants/day';
import { generateColor } from './color';
import { maxHoursPerDay } from './count';

export function getDocDefinition(timetable, ...args) {
  const content = Object.keys(timetable).map((timetableTitle, i) => {
    const timetableData = timetable[timetableTitle];

    return [
      buildTableTitle(timetableTitle, i),
      buildTable(timetableData, args),
    ];
  });

  return { content, pageOrientation: 'landscape' };
}

function buildTableTitle(timetableTitle, iteration) {
  const pageBreak = iteration === 0 ? null : 'before';

  return {
    text: timetableTitle,
    fontSize: 28,
    bold: true,
    margin: [0, 0, 0, 10],
    pageBreak,
  };
}

function buildTable(timetableData, args) {
  const widths = [30, '*', '*', '*', '*', '*'];
  const body = [
    buildBodyHeaders(),
    ...buildBodyRows(timetableData, args),
  ];

  return {
    table: { widths, body },
  };
}

function buildBodyHeaders() {
  return [
    { text: '#', alignment: 'center', bold: true },
    ...daysInWeek.map(day => ({ text: day, bold: true, margin: [3, 0] })),
  ];
}

function buildBodyRows(timetableData, args) {
  const hoursPerDay = maxHoursPerDay(timetableData);

  return Array(hoursPerDay)
    .fill()
    .map((_, hourIndex) => {
      return [
        { stack: ['\n', { text: hourIndex + 1, bold: true, alignment: 'center' }, '\n'], margin: [3, 4] },
        ...buildRowCells(timetableData, hourIndex, ...args),
      ];
    });
}

function buildRowCells(timetableData, hourIndex, getCardPrimaryText, getCardSecondaryText, monochromeMode) {
  return Array(daysInWeek.length)
    .fill()
    .map((_, dayIndex) => {
      if (!timetableData[dayIndex][hourIndex]) return { stack: [] };

      const subject = JSON.parse(timetableData[dayIndex][hourIndex]);
      return {
        stack: [{ text: getCardPrimaryText(subject), bold: true }, '\n', getCardSecondaryText(subject)],
        fillColor: monochromeMode ? null : generateColor(getCardPrimaryText(subject)),
        fillOpacity: 0.6,
        margin: [3, 4],
      };
    });
}

```

Reply:
I think it is mix of it being the end of project and coding style. I definitely agree this version is more readable.